File: packages/ios/ContentfulOptimization/Sources/ContentfulOptimization/Polyfills/NativePolyfills.swift
import Foundation
import JavaScriptCore
import os

/// Registers native Swift functions into a JSContext to back the JS polyfill scripts.
enum NativePolyfills {

    /// Per-context store for active timer work items, preventing cross-context collisions.
    final class TimerStore {
        private var timers: [Int: DispatchWorkItem] = [:]

        func set(_ id: Int, workItem: DispatchWorkItem) {
            timers[id] = workItem
        }

        func cancel(_ id: Int) {
            timers[id]?.cancel()
            timers.removeValue(forKey: id)
        }

        func fired(_ id: Int) {
            timers.removeValue(forKey: id)
        }

        /// Cancels all pending timers. Call when the owning context is destroyed.
        func cancelAll() {
            for (_, workItem) in timers {
                workItem.cancel()
            }
            timers.removeAll()
        }
    }

    /// Escapes a Swift string so it can be safely interpolated into a JS string literal.
    static func escapeForJS(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\\", with: "\\\\")
            .replacingOccurrences(of: "\"", with: "\\\"")
            .replacingOccurrences(of: "'", with: "\\'")
            .replacingOccurrences(of: "\n", with: "\\n")
            .replacingOccurrences(of: "\r", with: "\\r")
            .replacingOccurrences(of: "\t", with: "\\t")
    }

    private static let signpostLog = OSLog(
        subsystem: "com.contentful.optimization",
        category: "Performance"
    )

    /// Register all native polyfill functions into the given JSContext.
    ///
    /// Returns a ``TimerStore`` that the caller must retain for the lifetime of
    /// the context and call ``TimerStore/cancelAll()`` on teardown.
    @discardableResult
    static func register(
        in context: JSContext,
        logger: @escaping (String, String) -> Void
    ) -> TimerStore {
        let timerStore = TimerStore()
        weak var weakContext = context

        // MARK: - __nativeLog(level, msg)

        let nativeLog: @convention(block) (String, String) -> Void = { level, msg in
            logger(level, msg)
        }
        context.setObject(nativeLog, forKeyedSubscript: "__nativeLog" as NSString)

        // MARK: - __nativeSetTimeout(id, delayMs)

        let nativeSetTimeout: @convention(block) (Int, Int) -> Void = { timerId, delayMs in
            let workItem = DispatchWorkItem {
                guard let ctx = weakContext else { return }
                timerStore.fired(timerId)
                ctx.evaluateScript("__timerFired(\(timerId))")
            }
            timerStore.set(timerId, workItem: workItem)
            DispatchQueue.main.asyncAfter(
                deadline: .now() + .milliseconds(max(delayMs, 0)),
                execute: workItem
            )
        }
        context.setObject(nativeSetTimeout, forKeyedSubscript: "__nativeSetTimeout" as NSString)

        // MARK: - __nativeClearTimeout(id)

        let nativeClearTimeout: @convention(block) (Int) -> Void = { timerId in
            timerStore.cancel(timerId)
        }
        context.setObject(nativeClearTimeout, forKeyedSubscript: "__nativeClearTimeout" as NSString)

        // MARK: - __nativeRandomUUID()

        let nativeRandomUUID: @convention(block) () -> String = {
            UUID().uuidString.lowercased()
        }
        context.setObject(nativeRandomUUID, forKeyedSubscript: "__nativeRandomUUID" as NSString)

        // MARK: - __nativeFetch(url, method, headersJSON, body, callbackId)

        let nativeFetch: @convention(block) (String, String, String, JSValue, Int) -> Void = {
            urlString, method, headersJSON, bodyValue, callbackId in

            let diagLog = DiagnosticLogger.shared
            diagLog.debug("[fetch] \(method) \(urlString)")

            let log = NativePolyfills.signpostLog
            let fetchSignpostID = OSSignpostID(log: log)
            os_signpost(
                .begin, log: log, name: "Fetch Bridge Crossing",
                signpostID: fetchSignpostID, "%{public}s %{public}s", method, urlString
            )

            guard let url = Foundation.URL(string: urlString) else {
                diagLog.error("[fetch] Invalid URL: \(urlString)")
                DispatchQueue.main.async {
                    let escaped = NativePolyfills.escapeForJS(urlString)
                    weakContext?.evaluateScript(
                        "__fetchComplete(\(callbackId), 0, \"{}\", \"\", \"Invalid URL: \(escaped)\")"
                    )
                    os_signpost(
                        .end, log: log, name: "Fetch Bridge Crossing",
                        signpostID: fetchSignpostID, "error: invalid URL"
                    )
                }
                return
            }

            var request = URLRequest(url: url)
            request.httpMethod = method

            if let data = headersJSON.data(using: .utf8),
               let headers = try? JSONSerialization.jsonObject(with: data) as? [String: String]
            {
                for (key, value) in headers {
                    request.setValue(value, forHTTPHeaderField: key)
                }
            }

            if !bodyValue.isNull && !bodyValue.isUndefined, let bodyString = bodyValue.toString() {
                request.httpBody = bodyString.data(using: .utf8)
            }

            URLSession.shared.dataTask(with: request) { data, response, error in
                DispatchQueue.main.async {
                    guard let ctx = weakContext else {
                        diagLog.warning("[fetch] Context deallocated before response for \(urlString)")
                        os_signpost(
                            .end, log: log, name: "Fetch Bridge Crossing",
                            signpostID: fetchSignpostID, "context deallocated"
                        )
                        return
                    }

                    if let error = error {
                        diagLog.error("[fetch] Network error for \(urlString): \(error.localizedDescription)")
                        let escaped = NativePolyfills.escapeForJS(
                            error.localizedDescription
                        )
                        ctx.evaluateScript(
                            "__fetchComplete(\(callbackId), 0, \"{}\", \"\", \"\(escaped)\")"
                        )
                        os_signpost(
                            .end, log: log, name: "Fetch Bridge Crossing",
                            signpostID: fetchSignpostID, "error: %{public}s", escaped
                        )
                        return
                    }

                    let httpResponse = response as? HTTPURLResponse
                    let statusCode = httpResponse?.statusCode ?? 0
                    let bodySize = data?.count ?? 0
                    diagLog.debug("[fetch] Response \(statusCode) from \(urlString) (\(bodySize) bytes)")
                    if statusCode >= 400, let data = data, let body = String(data: data, encoding: .utf8) {
                        diagLog.error("[fetch] Error body: \(body)")
                    }

                    var responseHeaders: [String: String] = [:]
                    if let allHeaders = httpResponse?.allHeaderFields as? [String: String] {
                        responseHeaders = allHeaders
                    }
                    let headersJSONStr = (try? JSONSerialization.data(
                        withJSONObject: responseHeaders
                    )).flatMap { String(data: $0, encoding: .utf8) } ?? "{}"

                    let bodyText = data.flatMap { String(data: $0, encoding: .utf8) } ?? ""

                    let escapedBody = NativePolyfills.escapeForJS(bodyText)
                    let escapedHeaders = NativePolyfills.escapeForJS(headersJSONStr)

                    ctx.evaluateScript(
                        "__fetchComplete(\(callbackId), \(statusCode), \"\(escapedHeaders)\", \"\(escapedBody)\", \"\")"
                    )
                    os_signpost(
                        .end, log: log, name: "Fetch Bridge Crossing",
                        signpostID: fetchSignpostID, "status: %d", statusCode
                    )
                }
            }.resume()
        }
        context.setObject(nativeFetch, forKeyedSubscript: "__nativeFetch" as NSString)

        return timerStore
    }
}

File: packages/android/ContentfulOptimization/src/main/kotlin/com/contentful/optimization/polyfills/NativePolyfills.kt
package com.contentful.optimization.polyfills

import com.contentful.optimization.core.DiagnosticLogger
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import okhttp3.Call
import okhttp3.Callback
import okhttp3.MediaType.Companion.toMediaTypeOrNull
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import okhttp3.Response
import org.json.JSONObject
import java.io.IOException
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class TimerStore {
    private val timers = ConcurrentHashMap<Int, Job>()

    fun set(id: Int, job: Job) {
        timers[id] = job
    }

    fun cancel(id: Int) {
        timers[id]?.cancel()
        timers.remove(id)
    }

    fun fired(id: Int) {
        timers.remove(id)
    }

    fun cancelAll() {
        for ((_, job) in timers) {
            job.cancel()
        }
        timers.clear()
    }
}

fun escapeForJS(value: String): String =
    value
        .replace("\\", "\\\\")
        .replace("\"", "\\\"")
        .replace("'", "\\'")
        .replace("\n", "\\n")
        .replace("\r", "\\r")
        .replace("\t", "\\t")

class NativeImpl(
    private val scope: CoroutineScope,
    private val timerStore: TimerStore,
    private val evaluateJS: suspend (String) -> Unit,
    private val logger: (String, String) -> Unit,
) {
    private val okHttpClient = OkHttpClient()

    fun log(level: String, msg: String) {
        logger(level, msg)
    }

    fun setTimeout(id: Int, delayMs: Int) {
        val job = scope.launch {
            delay(delayMs.toLong().coerceAtLeast(0))
            timerStore.fired(id)
            evaluateJS("__timerFired($id)")
        }
        timerStore.set(id, job)
    }

    fun clearTimeout(id: Int) {
        timerStore.cancel(id)
    }

    fun randomUUID(): String = UUID.randomUUID().toString()

    fun fetch(url: String, method: String, headers: String, body: String, callbackId: Int) {
        DiagnosticLogger.debug { "[fetch] $method $url" }

        val requestBuilder = Request.Builder().url(url)

        try {
            val headersObj = JSONObject(headers)
            for (key in headersObj.keys()) {
                requestBuilder.addHeader(key, headersObj.getString(key))
            }
        } catch (_: Exception) {
            // headers was empty or invalid JSON — proceed without
        }

        val requestBody = if (body.isNotEmpty() && method != "GET" && method != "HEAD") {
            val contentType = "application/json".toMediaTypeOrNull()
            body.toRequestBody(contentType)
        } else {
            if (method == "POST" || method == "PUT" || method == "PATCH") {
                "".toRequestBody(null)
            } else {
                null
            }
        }

        val request = requestBuilder.method(method, requestBody).build()

        okHttpClient.newCall(request).enqueue(object : Callback {
            override fun onResponse(call: Call, response: Response) {
                val responseBody = response.body?.string() ?: ""
                val responseHeaders = JSONObject()
                for (name in response.headers.names()) {
                    responseHeaders.put(name, response.header(name))
                }
                val statusCode = response.code
                val bodySize = responseBody.length
                DiagnosticLogger.debug { "[fetch] Response $statusCode from $url ($bodySize bytes)" }
                if (statusCode >= 400) {
                    DiagnosticLogger.error { "[fetch] Error body: $responseBody" }
                }

                val escapedBody = escapeForJS(responseBody)
                val escapedHeaders = escapeForJS(responseHeaders.toString())

                scope.launch {
                    evaluateJS(
                        "__fetchComplete($callbackId, $statusCode, \"$escapedHeaders\", \"$escapedBody\", \"\")"
                    )
                }
            }

            override fun onFailure(call: Call, e: IOException) {
                DiagnosticLogger.error { "[fetch] Network error for $url: ${e.message}" }
                val escaped = escapeForJS(e.message ?: "Network error")
                scope.launch {
                    evaluateJS(
                        "__fetchComplete($callbackId, 0, \"{}\", \"\", \"$escaped\")"
                    )
                }
            }
        })
    }

    companion object {
        const val BOOTSTRAP_SCRIPT = """
            globalThis.__nativeLog          = function(l, m)        { return __native.log(l, m); };
            globalThis.__nativeSetTimeout   = function(id, ms)      { return __native.setTimeout(id, ms); };
            globalThis.__nativeClearTimeout = function(id)           { return __native.clearTimeout(id); };
            globalThis.__nativeRandomUUID   = function()             { return __native.randomUUID(); };
            globalThis.__nativeFetch        = function(u,m,h,b,cb)  { return __native.fetch(u, m, h, b, cb); };
        """
    }
}
